### PR TITLE
terminal: 'modifiable'; 'scrollback'; follow output only if cursor is on last line

### LINKED ISSUE #2637: Support for clearing scrollback with :term
This is an enhancement request for the (ahem, amazing) new `:term` feature.  I'd like a command that can clear a terminal's backscroll history similarly to tmux's `clear-history` command.

An example use-case from a typical workflow of mine:
1.   Make some project edits in nvim
2.   In a :term window, blow away the backscroll history
3.   Run a large test suite in the :term window
4.   Repeat the above steps.  A lot.

By clearing the backscroll in the test window (which often runs `vagrant ssh` into a project VM) before the run, it's much easier to jump to the beginning of the run, search through it, etc.  For example, I currently have bindings setup in tmux (`<prefix>k`) which clears both the shell window (via `<C-L>`) as well as the backscroll (via tmux `clear-history`).

### PR DESCRIPTION
**Pull request by justinmk**
Closes #2637. 

- Will crash if *all* lines are deleted. 
    - It's difficult to support line deletions, because scrollback is received passively from libvterm. Instead, one can safely and correctly "delete" lines by temporarily setting 'scrollback' option to a smaller value.

References #2607
References #2637
References #5431

### CODE REVIEW

File: runtime/doc/options.txt [hunk 1]
```diff
@@ -4949,6 +4949,15 @@ A jump table for the options with a short description can be found at |Q_op|.
 	be used as the new value for 'scroll'.  Reset to half the window
 	height with ":set scroll=0".
 
+			*'scrollback'* *'scbk'* *'noscrollback'* *'noscbk'*
+'scrollback' 'scbk'	boolean  (default: 1000
```
Comment (jamessan) [inline runtime/doc/options.txt]: This isn't a boolean option, so the "boolean" and "noscrollback"/"noscbk" parts should be removed.

File: runtime/doc/options.txt [hunk 2]
```diff
@@ -4949,6 +4949,15 @@ A jump table for the options with a short description can be found at |Q_op|.
...
+				  in non-terminal buffers: -1)
+			local to buffer
+	Maximum number of lines kept beyond the visible screen. Lines at the
+	top are deleted if new lines exceed this limit.
+	Only allowed in "terminal" buffers. |terminal-emulator| 'buftype'
```
Comment (ZyX-I) [inline runtime/doc/options.txt]: I do not think you meant `'buftype'` here.

**Comment by justinmk** [comment]
Followup PR on the way to further improve cursor behavior and 'cursorline'.

**Comment by fmoralesc** [comment]
I either messed up my build, or this PR was ineffective in setting 'modifiable' for terminal buffers :/

**Comment by justinmk** [comment]
@fmoralesc It's not modifiable by default, it now _allows_ `:set modifiable`.

Note there is a known issue: deleting lines may cause a crash. Deleting lines isn't supported (it should be prevented in a future PR).

**Comment by fmoralesc** [comment]
Gotcha. Modifiable terminals by default would be so useful, it would be nice if those issues could be ironed out.

**Comment by justinmk** [comment]
You mean, like being able to do Vim edits to the terminal content and then having those edits reflected in the child process? That's not really possible. 

`:set modifiable` is intended only for in-line edits, for special cases like easymotion. Someone also mentioned adding `{{{` markers for folding scenarios.

**Comment by fmoralesc** [comment]
No, treating the output of commands like `ls` as things you could filter with other commands in the editor, like in acme. Terminals could work as a sort of scratch buffer. 

I don't care about the edits being reflected in the child process, at all.

**Comment by justinmk** [comment]
We need feedback on that then. And we can try to improve it for that use case, based on feedback. 

But even after that, I'd still think 'nomodifiable' is advisable as the default. It's easy enough for users who want to filter terminal buffers to make 'modifiable' their default:

    au TermOpen * setlocal modifiable

**Comment by fmoralesc** [comment]
Sure. I thought you were aiming for making terminals modifiable thinking on precisely this use case, not as a workaround for how plugins like easymotion work... I guess I just saw what I wanted to see ;)

**Comment by justinmk** [comment]
The changes here are a necessary step in either case. After these changes, AFAIK you _can_ achieve what you describe above, (as long as you don't delete "too many" lines). 

OTOH wouldn't it be better if one could pipe `:terminal` output similar to `:!`? E.g.:

    :read :terminal ls

**Comment by fmoralesc** [comment]
For sure. As it is now, it works perfectly for what I describe. I even tried deleting all lines, and I didn't get any issues (when I entered terminal mode again, the cursor was reset by my shell).

> OTOH wouldn't it be better if one could pipe :terminal output

Dunno, it might be useful, but I'm not sure it would be better. I need to think more about this, might implement something as part of the acme plugin thingie I was working on...

**Comment by justinmk** [comment]
> I even tried deleting all lines, and I didn't get any issues (when I entered terminal mode again, the cursor was reset by my shell).

It's timing-dependent. However it "shouldn't" crash if you leave lines >= the window height.

**Comment by tecfu** [comment]
@justinmk Thanks for you hard work on this, Justin. Awesome.

**Comment by cmcaine** [comment]
@fmoralesc I don't understand the use case you're describing, but it sounds interesting :) Could you post a link?

**Pull request merged**